### PR TITLE
Refactor payloads

### PR DESCRIPTION
**Pull request by benpturner**
Re-wrote a large majority of the payloads.py so that it pulls the templates out to files so users can modify their own.

Also created a function to compile every *.c file in the payload-templates so users can create their own EXE templates and it will compile them all at startup or via createnewpayload command

### CODE REVIEW

File: poshc2/server/Payloads.py [hunk 1]
```diff
@@ -27,6 +27,7 @@ def __init__(self, KillDate, Key, Insecure, UserAgent, Referrer, ConnectURL, Bas
         self.KillDate = KillDate
         self.Key = Key
         self.QuickCommand = select_item("QuickCommand", "C2Server")
+        self.FirstURL = get_first_url(select_item("PayloadCommsHost", "C2Server"), select_item("DomainFrontHeader", "C2Server"))
```
Comment (riskydissonance) [inline poshc2/server/Payloads.py:30]: Longer term we should split the array out when we create the C2 server (or a new payload) and add a new row in the URLs table for each one, then for this stuff you can use `DB.get_default_url_id()`.

Would then allow you use `show-urls` to see all the urls, and can set the failover etc just using IDs or names from the list.

File: poshc2/server/Payloads.py [hunk 2]
```diff
@@ -313,361 +312,308 @@ def CreateShellcode(self, name=""):
...
-</scRIPt></sCrIptlEt>
-""" % basefile
-        filename = "%srg_sct.xml" % (self.BaseDirectory)
-        output_file = open(filename, 'w')
-        output_file.write(raw1)
-        filename = "%scs_sct.xml" % (self.BaseDirectory)
-        output_file.close()
-        output_file = open(filename, 'w')
-        output_file.write(raw2)
+    def CreateSCT(self, name=""):
```
Comment (riskydissonance) [inline poshc2/server/Payloads.py:290]: This is so much cleaner :ok_hand:

File: poshc2/server/Payloads.py [hunk 3]
```diff
@@ -313,361 +312,308 @@ def CreateShellcode(self, name=""):
...
-        with open("%sDotNet2JS.js" % PayloadTemplatesDirectory, 'r') as f:
-            dotnet = f.read()        
-        with open('%s%sPosh_v4_x64_Shellcode.b64' % (self.BaseDirectory, name), 'rb') as f:
-            v4_64 = f.read()
-        with open('%s%sPosh_v4_x86_Shellcode.b64' % (self.BaseDirectory, name), 'rb') as f:
-            v4_86 = f.read()
-        dotnet = dotnet.replace("#REPLACEME32#", v4_86.decode('utf-8'))  
-        dotnet = dotnet.replace("#REPLACEME64#", v4_64.decode('utf-8'))        
+
+        if os.name == 'nt':
```
Comment (riskydissonance) [inline poshc2/server/Payloads.py]: I don't think we actually need this anymore as we don't support running natively on Windows - only using Docker

Comment (benpturner) [reply poshc2/server/Payloads.py]: Yeah that's a good point actually!

File: poshc2/server/Payloads.py [hunk 4]
```diff
@@ -313,361 +312,308 @@ def CreateShellcode(self, name=""):
...
-            if os.name == 'nt':
-                compile64 = "C:\\TDM-GCC-64\\bin\\gcc.exe %s%sPosh64_migrate.c -o %s%sPosh64_migrate.exe" % (self.BaseDirectory, name, self.BaseDirectory, name)
-                compile32 = "C:\\TDM-GCC-32\\bin\\gcc.exe %s%sPosh32_migrate.c -o %s%sPosh32_migrate.exe" % (self.BaseDirectory, name, self.BaseDirectory, name)
-            else:
-                compile64 = "x86_64-w64-mingw32-gcc -w %s%sPosh64_migrate.c -o %s%sPosh64_migrate.exe" % (self.BaseDirectory, name, self.BaseDirectory, name)
-                compile32 = "i686-w64-mingw32-gcc -w %s%sPosh32_migrate.c -o %s%sPosh32_migrate.exe" % (self.BaseDirectory, name, self.BaseDirectory, name)
-            subprocess.check_output(compile64, shell=True)
-            subprocess.check_output(compile32, shell=True)
+        # work out the compiler per OS
+        if os.name == 'nt':
```
Comment (riskydissonance) [inline poshc2/server/Payloads.py]: Same here with the Windows stuff

Comment (benpturner) [reply poshc2/server/Payloads.py]: Yeah, just removed to all places. Nice spot

File: poshc2/server/Payloads.py (hunk 1 above)
Comment (benpturner) [reply poshc2/server/Payloads.py:30]: Yeah we can update that going forward definitely now that multiple URLs have been added

**Review by riskydissonance — APPROVED** [review]
:+1: looks good!

**Pull request merged**